pkgs/development/tools/godot/default.nix

```diff
@@ -65,39 +83,76 @@ stdenv.mkDerivation rec {
 
   enableParallelBuilding = true;
 
-  sconsFlags = "target=release_debug platform=x11";
-  preConfigure = ''
-    sconsFlags+=" ${
-      lib.concatStringsSep " "
-      (lib.mapAttrsToList (k: v: "${k}=${builtins.toJSON v}") options)
-    }"
-  '';
+  sconsFlags = let toSconsValue = x: if builtins.isBool x then lib.boolToString x else x; in
+    lib.pipe self.passthru.sconsFlagSet [
+      (lib.mapAttrsToList (k: v: "${k}=${toSconsValue v}"))
+      (lib.concatStringsSep " ")
+    ];
 
   outputs = [ "out" "dev" "man" ];
 
+  installSteps = [ "installBinStep" "installDevStep" "installManStep" "installShortcutStep" ];
+
   installPhase = ''
-    mkdir -p "$out/bin"
-    cp bin/godot.* $out/bin/godot
+    runHook preInstall
+
+    for curInstallStep in $installSteps; do
+      eval "''${!curInstallStep}"
+    done
+
+    runHook postInstall
+  '';
+
+  installBinStep = ''
+    runHook preInstallBin
+
+    bin="$out"/"$binInstallPath"
+    mkdir -p "$bin"
+    cp -R bin/. "$bin"
+    mv "$bin"/$builtGodotBinNamePattern "$bin/$installedGodotBinName"
```

Comment:
This is defined in nix. We should use the nix variables.

Reply:
The fact that `binInstallPath`, `builtGodotBinNamePattern`, and `installedGodotBinName` are defined outside of passthru makes them accessible within the phases as environment variables.  I _could_ rewrite these as:
```sh
bin="$out"/"${self.binInstallPath}"
mkdir -p "$bin"
cp -R bin/. "$bin"
mv "$bin"/${self.builtGodotBinNamePattern} "$bin/${self.installedGodotBinName}"
```
but I don't see any value in doing so.

Reply:
> defined outside of passthru makes them accessible within the phases as environment variables.

No, not really. Either `rec` could be used or they are moved into inputs which is the cheapest of the three.

Reply:
> No, not really.

Are you saying that having them defined as attributes in the attrset given to mkDerivation does _not_ make them accessible as environment variables within the phases?  Why is it, then, that these derivations actually build, and that when you go to a `nix-shell` or `nix develop` shell, `$binInstallPath` and such resolve?

What what would `rec` accomplish?  Also, could you provide a concrete example of how you would pass these to "inputs"?  I'm guessing you mean `nativeBuildInputs` but I don't know what would happen if you put a string, as opposed to a derivation, in that list.

As for the quoting pattern for `bin`, sure I also prefer putting the entire path in quotes, rather than separately quoting out and binInstallPath.  That was an oversight.  I'll correct it.